src/testutils.rs

```diff
@@ -0,0 +1,124 @@
+use std::io;
+
+use async_trait::async_trait;
+use tokio::sync::{mpsc, watch};
+
+use crate::cid::{ConnectionId, ConnectionPeer};
+use crate::udp::AsyncUdpSocket;
+
+/// A mock socket that can be used to simulate a perfect link.
+#[derive(Debug)]
+pub struct MockUdpSocket {
+    outbound: mpsc::UnboundedSender<Vec<u8>>,
+    inbound: mpsc::UnboundedReceiver<Vec<u8>>,
+    /// Peers identified by a letter
+    pub only_peer: char,
+    /// Watch for signal that the link is up. If not up, link will SILENTLY drop all sent packets.
+    /// If up_status is None, the link is always up.
+    up_status: Option<watch::Receiver<bool>>,
+}
+
+impl MockUdpSocket {
+    /// Grab the status controller for this socket.
+    /// Can only be taken once, otherwise fails.
+    pub fn take_status_controller(&mut self) -> Result<watch::Sender<bool>, String> {
+        match self.up_status {
+            None => {
+                let (up_status_tx, up_status) = watch::channel(true);
+                self.up_status = Some(up_status);
+                Ok(up_status_tx)
+            }
+            Some(_) => Err("Status controller already taken".to_string()),
+        }
+    }
+
+    /// Should the link send packets?
+    fn active(&self) -> bool {
+        match &self.up_status {
+            None => true,
+            Some(up_status) => *up_status.borrow(),
+        }
+    }
+}
+
+#[async_trait]
+impl AsyncUdpSocket<char> for MockUdpSocket {
+    async fn send_to(&mut self, buf: &[u8], target: &char) -> io::Result<usize> {
+        if target != &self.only_peer {
+            panic!("MockUdpSocket only supports sending to one peer");
+        }
+        if !self.active() {
+            return Ok(buf.len());
+        }
+        let result = self.outbound.send(buf.to_vec());
+        if let Err(err) = result {
+            Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                format!("channel closed: {err}"),
+            ))
+        } else {
+            Ok(buf.len())
+        }
+    }
+
+    async fn recv_from(&mut self, buf: &mut [u8]) -> io::Result<(usize, char)> {
+        let packet = self
+            .inbound
+            .recv()
+            .await
+            .ok_or_else(|| io::Error::new(io::ErrorKind::UnexpectedEof, "channel closed"))?;
+        if buf.len() < packet.len() {
+            panic!("buffer too small for perfect link");
```

Comment:
do we need to panic here?

Reply:
So the other options I see:
- Return n=0, which gives the tester a warning log "unable to decode uTP packet" 
- Put the first part of the packet, up to the buffer size, into the buffer. again, giving an "unable to decode uTP packet"
- Return an `io::Error`, maybe an `InvalidInput`. Note that src/socket seems to drop errors on the floor without a log

Any other suggestions?

Since this is a utility aimed at testing instead of production, panicking with a clear reason is more important than staying alive. None of these other options seem to give the tester much information about why it failed, and some will failing quietly, which means a long timeout/retry loop that's hard to debug.

So I don't see a better option than panicking, yet.

Reply:
fair enough. I'm okay with keeping the panic, but we should document it as we did above for `send_to`.